Mathlib/CategoryTheory/Monad/Monadicity.lean

```diff
@@ -0,0 +1,490 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+
+! This file was ported from Lean 3 source module category_theory.monad.monadicity
+! leanprover-community/mathlib commit 4bd8c855d6ba8f0d5eefbf80c20fa00ee034dec9
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.CategoryTheory.Limits.Preserves.Shapes.Equalizers
+import Mathlib.CategoryTheory.Limits.Shapes.Reflexive
+import Mathlib.CategoryTheory.Monad.Coequalizer
+import Mathlib.CategoryTheory.Monad.Limits
+
+/-!
+# Monadicity theorems
+
+We prove monadicity theorems which can establish a given functor is monadic. In particular, we
+show three versions of Beck's monadicity theorem, and the reflexive (crude) monadicity theorem:
+
+`G` is a monadic right adjoint if it has a right adjoint, and:
+
+* `D` has, `G` preserves and reflects `G`-split coequalizers, see
+  `category_theory.monad.monadic_of_has_preserves_reflects_G_split_coequalizers`
+* `G` creates `G`-split coequalizers, see
+  `category_theory.monad.monadic_of_creates_G_split_coequalizers`
+  (The converse of this is also shown, see
+   `category_theory.monad.creates_G_split_coequalizers_of_monadic`)
+* `D` has and `G` preserves `G`-split coequalizers, and `G` reflects isomorphisms, see
+  `category_theory.monad.monadic_of_has_preserves_G_split_coequalizers_of_reflects_isomorphisms`
+* `D` has and `G` preserves reflexive coequalizers, and `G` reflects isomorphisms, see
+  `category_theory.monad.monadic_of_has_preserves_reflexive_coequalizers_of_reflects_isomorphisms`
+
+## Tags
+
+Beck, monadicity, descent
+
+## TODO
+
+Dualise to show comonadicity theorems.
+-/
+
+-- Porting note: TODO remove
+set_option autoImplicit false
+
+universe v₁ v₂ u₁ u₂
+
+namespace CategoryTheory
+
+namespace Monad
+
+open Limits
+
+noncomputable section
+
+-- Hide the implementation details in this namespace.
+namespace MonadicityInternal
+
+section
+
+variable {C : Type u₁} {D : Type u₂}
+
+variable [Category.{v₁} C] [Category.{v₁} D]
+
+variable {G : D ⥤ C} [IsRightAdjoint G]
+
+-- Porting note: these were `local notation` in mathlib3. However, `abbrev` is not always
+-- sufficient for synthesizing the right instances, so some occurrences have been inlined as a
+-- temporary measure. All occurrences of `leftAdjoint G` and `Adjunction.ofRightAdjoint G` are the
+-- result of inlining.
+abbrev F := leftAdjoint G
+abbrev adj := Adjunction.ofRightAdjoint G
```

Comment:
@kmill, do you have any suggestions for handling the notation here?

Reply:
It's probably `local notation3 "F" => leftAdjoint G` and `local notation3 "adj" => Adjunction.ofRightAdjoint G`.

Reply:
Should I be able to dot off this? eg `F.obj`?

Reply:
You should be able to do `(F).map` for example

Reply:
Any way to avoid that?

Reply:
Forgot to turn off hygiene. Have `notation3` now

Reply:
My suggestion was `local notation3 "F" => leftAdjoint G` (with a `3` in there). This does not need the unhygienic option, and it actually pretty prints. For this module though, you need a couple copies of the notation because `G` gets introduced multiple times. #5826

(Regarding if there's a way to avoid parentheses, if you check out that link in my comment, you'll see there was a similar problem. The Lean 3 workaround was `F .map`, but that doesn't work anymore.)

Reply:
Sorry. It got a little late for me. I tried a few things, got something to work, and misremembered what I did.